File: ios/PSMobile/Testkoerper.swift
import Foundation
import simd

/// Ein Würfel als STL, gerechnet statt mitgeliefert.
///
/// Eine Datei im Bundle wäre in jeder ausgelieferten App dabei, nur
/// damit ein Test etwas zum Anfassen hat.
///
/// Jede Zahl wird einzeln angehängt. Der erste Anlauf schrieb die zwölf
/// Fließkommazahlen eines Dreiecks aus einem zusammengesetzten Array —
/// dabei kamen nur sechs an, und die Datei war 396 statt 684 Bytes groß.
/// libslic3r meldete daraufhin nur „Loading of a model file failed",
/// ohne zu sagen, woran.
enum Testkoerper {

    /// Binäres STL eines achsenparallelen Würfels mit der Kante `kante`.
    static func wuerfel(kante a: Float = 20) -> Data {
        let ecken: [SIMD3<Float>] = [
            [0, 0, 0], [a, 0, 0], [a, a, 0], [0, a, 0],
            [0, 0, a], [a, 0, a], [a, a, a], [0, a, a],
        ]
        // Zwoelf Dreiecke, von aussen gesehen gegen den Uhrzeigersinn.
        let flaechen: [(Int, Int, Int)] = [
            (0, 2, 1), (0, 3, 2),   // unten
            (4, 5, 6), (4, 6, 7),   // oben
            (0, 1, 5), (0, 5, 4),   // vorn
            (1, 2, 6), (1, 6, 5),   // rechts
            (2, 3, 7), (2, 7, 6),   // hinten
            (3, 0, 4), (3, 4, 7),   // links
        ]

        // 80 Byte Kopf, vier Byte Anzahl, dann je Dreieck 50 Byte.
        var daten = Data(count: 80)
        func zahl(_ wert: Float) {
            var f = wert
            withUnsafeBytes(of: &f) { daten.append(contentsOf: $0) }
        }
        var anzahl = UInt32(flaechen.count)
        withUnsafeBytes(of: &anzahl) { daten.append(contentsOf: $0) }

        for (i, j, k) in flaechen {
            // Die Normale darf null bleiben - libslic3r rechnet sie aus
            // der Reihenfolge der Ecken.
            zahl(0); zahl(0); zahl(0)
            for ecke in [ecken[i], ecken[j], ecken[k]] {
                zahl(ecke.x); zahl(ecke.y); zahl(ecke.z)
            }
            var attribut = UInt16(0)
            withUnsafeBytes(of: &attribut) { daten.append(contentsOf: $0) }
        }
        return daten
    }

    /// Schreibt den Würfel in eine Datei und gibt ihren Pfad zurück.
    static func wuerfelDatei(kante: Float = 20,
                             name: String = "psm-testwuerfel.stl") throws -> URL {
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(name)
        try wuerfel(kante: kante).write(to: url)
        return url
    }

    /// Ein Kegel mit vielen Facetten, gerechnet wie der Wuerfel.
    ///
    /// Fuer die Untersuchung der G-Code-Vorschau in dieser Nachtsitzung:
    /// ein einzelner Wuerfel hat nur wenige, kurze Perimeter und keine
    /// Ueberhaenge - er stellt weder Stuetzen noch viele Schichten noch
    /// gekruemmte Konturen auf die Probe. Ein hoher, feinfacettierter
    /// Kegel schon.
    static func kegel(radius r: Float = 30, hoehe h: Float = 60,
                       seiten n: Int = 64) -> Data {
        var ecken: [SIMD3<Float>] = []
        for i in 0..<n {
            let winkel = Float(i) / Float(n) * 2 * .pi
            ecken.append([r * cos(winkel), r * sin(winkel), 0])
        }
        let spitze = SIMD3<Float>(0, 0, h)
        let mitte = SIMD3<Float>(0, 0, 0)

        var daten = Data(count: 80)
        func zahl(_ wert: Float) {
            var f = wert
            withUnsafeBytes(of: &f) { daten.append(contentsOf: $0) }
        }
        func dreieck(_ a: SIMD3<Float>, _ b: SIMD3<Float>, _ c: SIMD3<Float>) {
            zahl(0); zahl(0); zahl(0)
            for ecke in [a, b, c] {
                zahl(ecke.x); zahl(ecke.y); zahl(ecke.z)
            }
            var attribut = UInt16(0)
            withUnsafeBytes(of: &attribut) { daten.append(contentsOf: $0) }
        }

        let anzahlDreiecke = UInt32(n * 2)
        var anzahl = anzahlDreiecke
        withUnsafeBytes(of: &anzahl) { daten.append(contentsOf: $0) }

        for i in 0..<n {
            let j = (i + 1) % n
            dreieck(mitte, ecken[j], ecken[i])       // Boden
            dreieck(ecken[i], ecken[j], spitze)      // Mantel
        }
        return daten
    }

    static func kegelDatei(name: String = "psm-testkegel.stl") throws -> URL {
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(name)
        try kegel().write(to: url)
        return url
    }
}

File: android/app/src/main/java/de/psmobile/diagnose/Testkoerper.kt
package de.psmobile.diagnose

import android.content.Context
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Der Wuerfel, mit dem der Selbsttest arbeitet.
 *
 * Erzeugt und nicht mitgeliefert: eine Datei in den Assets waere in
 * jedem APK, obwohl sie nur beim Selbsttest gebraucht wird, und ein
 * Modell aus dem Dateisystem des Nutzers waere kein Test, sondern ein
 * Zufall. Zwoelf Dreiecke sind schnell geschrieben und immer gleich -
 * genau das macht die Messwerte zwischen zwei Geraeten vergleichbar.
 *
 * Binaeres STL: 80 Byte Kopf, 4 Byte Anzahl, dann je 50 Byte pro
 * Dreieck (Normale und drei Ecken als float32, dazu zwei ungenutzte
 * Byte). Little Endian, so will es das Format.
 */
object Testkoerper {

    /** Kantenlaenge in Millimetern. Gross genug fuer echte Schichten. */
    private const val KANTE = 20f

    /**
     * Der Dateiname ist derselbe wie auf iOS (`Testkoerper.wuerfelDatei`,
     * `name: String = "psm-testwuerfel.stl"`). Er ist sichtbar: der Kern
     * uebernimmt den Dateinamen als Objektnamen, und der steht in der
     * Modell-Liste. Bis zum 10.09.2026 hiess er hier "wuerfel.stl" -
     * derselbe Rundgang zeigte auf beiden Geraeten einen anderen Namen.
     */
    fun wuerfelDatei(context: Context, name: String = "psm-testwuerfel.stl"): File {
        val ordner = File(context.cacheDir, "selbsttest").also { it.mkdirs() }
        return File(ordner, name).also { if (!it.exists()) it.writeBytes(wuerfel()) }
    }

    private fun wuerfel(): ByteArray {
        val k = KANTE
        // Acht Ecken, Ursprung in einer unteren Kante.
        val e = arrayOf(
            floatArrayOf(0f, 0f, 0f), floatArrayOf(k, 0f, 0f),
            floatArrayOf(k, k, 0f), floatArrayOf(0f, k, 0f),
            floatArrayOf(0f, 0f, k), floatArrayOf(k, 0f, k),
            floatArrayOf(k, k, k), floatArrayOf(0f, k, k),
        )
        // Zwoelf Dreiecke, jede Flaeche zweimal geteilt. Die Reihenfolge
        // der Ecken bestimmt, wohin die Normale zeigt - nach aussen, sonst
        // haelt der Slicer den Koerper fuer umgestuelpt.
        val dreiecke = listOf(
            intArrayOf(0, 3, 2), intArrayOf(0, 2, 1),   // unten
            intArrayOf(4, 5, 6), intArrayOf(4, 6, 7),   // oben
            intArrayOf(0, 1, 5), intArrayOf(0, 5, 4),   // vorn
            intArrayOf(2, 3, 7), intArrayOf(2, 7, 6),   // hinten
            intArrayOf(1, 2, 6), intArrayOf(1, 6, 5),   // rechts
            intArrayOf(3, 0, 4), intArrayOf(3, 4, 7),   // links
        )

        val puffer = ByteBuffer.allocate(84 + dreiecke.size * 50)
            .order(ByteOrder.LITTLE_ENDIAN)
        puffer.put(ByteArray(80))                 // Kopf, Inhalt egal
        puffer.putInt(dreiecke.size)

        dreiecke.forEach { d ->
            val a = e[d[0]]; val b = e[d[1]]; val c = e[d[2]]
            normale(a, b, c).forEach { puffer.putFloat(it) }
            listOf(a, b, c).forEach { p -> p.forEach { puffer.putFloat(it) } }
            puffer.putShort(0)                    // Attributbytes, ungenutzt
        }
        return puffer.array()
    }

    /**
     * Ein Kegel mit vielen Facetten, gerechnet wie der Wuerfel -
     * Gegenstueck zu `Testkoerper.kegel` in `ios/PSMobile/Testkoerper.swift`.
     *
     * Fuer die Untersuchung der G-Code-Vorschau: ein einzelner Wuerfel
     * hat nur wenige, kurze Perimeter und keine Ueberhaenge - er stellt
     * weder Stuetzen noch viele Schichten noch gekruemmte Konturen auf
     * die Probe. Ein hoher, feinfacettierter Kegel schon.
     */
    private fun kegel(r: Float = 30f, h: Float = 60f, n: Int = 64): ByteArray {
        val ecken = Array(n) { i ->
            val winkel = i.toFloat() / n * 2f * kotlin.math.PI.toFloat()
            floatArrayOf(r * kotlin.math.cos(winkel), r * kotlin.math.sin(winkel), 0f)
        }
        val spitze = floatArrayOf(0f, 0f, h)
        val mitte = floatArrayOf(0f, 0f, 0f)

        val puffer = ByteBuffer.allocate(84 + n * 2 * 50).order(ByteOrder.LITTLE_ENDIAN)
        puffer.put(ByteArray(80))
        puffer.putInt(n * 2)

        fun dreieck(a: FloatArray, b: FloatArray, c: FloatArray) {
            // Wie drueben: Normale null lassen, der Slicer rechnet sie neu.
            puffer.putFloat(0f); puffer.putFloat(0f); puffer.putFloat(0f)
            listOf(a, b, c).forEach { p -> p.forEach { puffer.putFloat(it) } }
            puffer.putShort(0)
        }
        for (i in 0 until n) {
            val j = (i + 1) % n
            dreieck(mitte, ecken[j], ecken[i])      // Boden
            dreieck(ecken[i], ecken[j], spitze)     // Mantel
        }
        return puffer.array()
    }

    /**
     * Die mitgelieferte Schraube (STEP) aus den Assets - fuer den Randfall
     * "STEP laedt", wenn der Test keinen lesbaren Pfad mitgibt. Gegenstueck:
     * `Bundle.main.path(forResource: "screw", ofType: "step")` drueben.
     */
    fun stepDatei(context: Context, name: String = "screw.step"): File {
        val ordner = File(context.cacheDir, "selbsttest").also { it.mkdirs() }
        return File(ordner, name).also { ziel ->
            if (!ziel.exists()) context.assets.open("selbsttest/$name").use { quelle ->
                ziel.outputStream().use { quelle.copyTo(it) }
            }
        }
    }

    fun kegelDatei(context: Context, name: String = "psm-testkegel.stl"): File {
        val ordner = File(context.cacheDir, "selbsttest").also { it.mkdirs() }
        return File(ordner, name).also { if (!it.exists()) it.writeBytes(kegel()) }
    }

    /** Kreuzprodukt der beiden Kanten, auf Laenge eins gebracht. */
    private fun normale(a: FloatArray, b: FloatArray, c: FloatArray): FloatArray {
        val u = floatArrayOf(b[0] - a[0], b[1] - a[1], b[2] - a[2])
        val v = floatArrayOf(c[0] - a[0], c[1] - a[1], c[2] - a[2])
        val n = floatArrayOf(
            u[1] * v[2] - u[2] * v[1],
            u[2] * v[0] - u[0] * v[2],
            u[0] * v[1] - u[1] * v[0],
        )
        val laenge = kotlin.math.sqrt(n[0] * n[0] + n[1] * n[1] + n[2] * n[2])
        return if (laenge == 0f) n else floatArrayOf(n[0] / laenge, n[1] / laenge, n[2] / laenge)
    }
}
